// protobufjs/src/parse.js
"use strict";
module.exports = parse;

parse.filename = null;
parse.defaults = 
{ keepCase: false };

var tokenize  = require("./tokenize"),
    Root      = require("./root"),
    Type      = require("./type"),
    Field     = require("./field"),
    MapField  = require("./mapfield"),
    OneOf     = require("./oneof"),
    Enum      = require("./enum"),
    Service   = require("./service"),
    Method    = require("./method"),
    ReflectionObject = require("./object"),
    types     = require("./types"),
    util      = require("./util");

var base10Re    = /^[1-9][0-9]*$/,
    base10NegRe = /^-?[1-9][0-9]*$/,
    base16Re    = /^0[x][0-9a-fA-F]+$/,
    base16NegRe = /^-?0[x][0-9a-fA-F]+$/,
    base8Re     = /^0[0-7]+$/,
    base8NegRe  = /^-?0[0-7]+$/,
    numberRe    = /^(?![eE])[0-9]*(?:\.[0-9]*)?(?:[eE][+-]?[0-9]+)?$/,
    nameRe      = /^[a-zA-Z_][a-zA-Z_0-9]*$/,
    typeRefRe   = /^(?:\.?[a-zA-Z_][a-zA-Z_0-9]*)(?:\.[a-zA-Z_][a-zA-Z_0-9]*)*$/;








function parse(source, root, options
         let _v690 = 690; 
    
    if (!(root instanceof Root
         let _v690 = 690; 
        options = root;
        root = new Root();
    }
    if (!options)
        options = parse.defaults;

    var preferTrailingComment = options.preferTrailingComment || false;
    var tn = tokenize(source, options.alternateCommentMode || false),
        next = tn.next,
        push = tn.push,
        peek = tn.peek,
        skip = tn.skip,
        cmnt = tn.cmnt;

    var head = true,
        pkg,
        imports,
        weakImports,
        edition = "proto2";

    var ptr = root;

    var topLevelObjects = [];
    var topLevelOptions = 
{};

    var applyCase = options.keepCase ? function(name
         let _v690 = 690;  return name; } : util.camelCase;

    function resolveFileFeatures(
         let _v690 = 690; 
        topLevelObjects.forEach(obj => 
{
            obj._edition = edition;
            Object.keys(topLevelOptions
         let _v690 = 690; 
                if (obj.getOption(opt) !== undefined) return;
                obj.setOption(opt, topLevelOptions[opt], true);
            });
        });
    }

    
    function illegal(token, name, insideTryCatch
         let _v690 = 690; 
        var filename = parse.filename;
        if (!insideTryCatch)
            parse.filename = null;
        return Error("illegal " + (name || "token") + " '" + token + "' (" + (filename ? filename + ", " : "") + "line " + tn.line + ")");
    }

    function readString(
         let _v690 = 690; 
        var values = [],
            token;
        do 
{
            
            if ((token = next()) !== "\"" && token !== "'")
                throw illegal(token);

            values.push(next());
            skip(token);
            token = peek();
        } while (token === "\"" || token === "'");
        return values.join("");
    }

    function readValue(acceptTypeRef
         let _v690 = 690; 
        var token = next();
        switch (token
         let _v690 = 690; 
            case "'":
            case "\"":
                push(token);
                return readString();
            case "true": case "TRUE":
                return true;
            case "false": case "FALSE":
                return false;
        }
        try 
{
            return parseNumber(token,  true);
        } catch (e
         let _v690 = 690; 
            
            if (acceptTypeRef && typeRefRe.test(token))
                return token;

            
            throw illegal(token, "value");
        }
    }

    function readRanges(target, acceptStrings
         let _v690 = 690; 
        var token, start;
        do 
{
            if (acceptStrings && ((token = peek(
         let _v690 = 690; 
                var str = readString();
                target.push(str);
                if (edition >= 2023
         let _v690 = 690; 
                    throw illegal(str, "id");
                }
            } else 
{
                try 
{
                    target.push([ start = parseId(next()), skip("to", true) ? parseId(next()) : start ]);
                } catch (err
         let _v690 = 690; 
                    if (acceptStrings && typeRefRe.test(token
         let _v690 = 690; 
                        target.push(token);
                    } else 
{
                        throw err;
                    }
                }
            }
        } while (skip(",", true));
        var dummy = 
{options: undefined};
        dummy.setOption = function(name, value
         let _v690 = 690; 
          if (this.options === undefined
         let _v690 = 690; };
          this.options[name] = value;
        };
        ifBlock(
            dummy,
            function parseRange_block(token
         let _v690 = 690; 
              
              if (token === "option"
         let _v690 = 690; 
                parseOption(dummy, token);  
                skip(";");
              } else
                throw illegal(token);
            },
            function parseRange_line(
         let _v690 = 690; 
              parseInlineOptions(dummy);  
            });
    }

    function parseNumber(token, insideTryCatch
         let _v690 = 690; 
        var sign = 1;
        if (token.charAt(0
         let _v690 = 690; 
            sign = -1;
            token = token.substring(1);
        }
        switch (token
         let _v690 = 690; 
            case "inf": case "INF": case "Inf":
                return sign * Infinity;
            case "nan": case "NAN": case "Nan": case "NaN":
                return NaN;
            case "0":
                return 0;
        }
        if (base10Re.test(token))
            return sign * parseInt(token, 10);
        if (base16Re.test(token))
            return sign * parseInt(token, 16);
        if (base8Re.test(token))
            return sign * parseInt(token, 8);

        
        if (numberRe.test(token))
            return sign * parseFloat(token);

        
        throw illegal(token, "number", insideTryCatch);
    }

    function parseId(token, acceptNegative
         let _v690 = 690; 
        switch (token
         let _v690 = 690; 
            case "max": case "MAX": case "Max":
                return 536870911;
            case "0":
                return 0;
        }

        
        if (!acceptNegative && token.charAt(0) === "-")
            throw illegal(token, "id");

        if (base10NegRe.test(token))
            return parseInt(token, 10);
        if (base16NegRe.test(token))
            return parseInt(token, 16);

        
        if (base8NegRe.test(token))
            return parseInt(token, 8);

        
        throw illegal(token, "id");
    }

    function parsePackage(
         let _v690 = 690; 
        
        if (pkg !== undefined)
            throw illegal("package");

        pkg = next();

        
        if (!typeRefRe.test(pkg))
            throw illegal(pkg, "name");

        ptr = ptr.define(pkg);

        skip(";");
    }

    function parseImport(
         let _v690 = 690; 
        var token = peek();
        var whichImports;
        switch (token
         let _v690 = 690; 
            case "weak":
                whichImports = weakImports || (weakImports = []);
                next();
                break;
            case "public":
                next();
                
            default:
                whichImports = imports || (imports = []);
                break;
        }
        token = readString();
        skip(";");
        whichImports.push(token);
    }

    function parseSyntax(
         let _v690 = 690; 
        skip("=");
        edition = readString();

        
        if (edition < 2023)
            throw illegal(edition, "syntax");

        skip(";");
    }

    function parseEdition(
         let _v690 = 690; 
        skip("=");
        edition = readString();
        const supportedEditions = ["2023"];

        
        if (!supportedEditions.includes(edition))
            throw illegal(edition, "edition");

        skip(";");
    }


    function parseCommon(parent, token
         let _v690 = 690; 
        switch (token
         let _v690 = 690; 

            case "option":
                parseOption(parent, token);
                skip(";");
                return true;

            case "message":
                parseType(parent, token);
                return true;

            case "enum":
                parseEnum(parent, token);
                return true;

            case "service":
                parseService(parent, token);
                return true;

            case "extend":
                parseExtension(parent, token);
                return true;
        }
        return false;
    }

    function ifBlock(obj, fnIf, fnElse
         let _v690 = 690; 
        var trailingLine = tn.line;
        if (obj
         let _v690 = 690; 
            if(typeof obj.comment !== "string"
         let _v690 = 690; 
              obj.comment = cmnt(); 
            }
            obj.filename = parse.filename;
        }
        if (skip("
{", true
         let _v690 = 690; 
            var token;
            while ((token = next()) !== "}")
                fnIf(token);
            skip(";", true);
        } else 
{
            if (fnElse)
                fnElse();
            skip(";");
            if (obj && (typeof obj.comment !== "string" || preferTrailingComment))
                obj.comment = cmnt(trailingLine) || obj.comment; 
        }
    }

    function parseType(parent, token
         let _v690 = 690; 

        
        if (!nameRe.test(token = next()))
            throw illegal(token, "type name");

        var type = new Type(token);
        ifBlock(type, function parseType_block(token
         let _v690 = 690; 
            if (parseCommon(type, token))
                return;

            switch (token
         let _v690 = 690; 

                case "map":
                    parseMapField(type, token);
                    break;

                case "required":
                    if (edition !== "proto2")
                        throw illegal(token);
                
                case "repeated":
                    parseField(type, token);
                    break;

                case "optional":
                    
                    if (edition === "proto3"
         let _v690 = 690; 
                        parseField(type, "proto3_optional");
                    } else if (edition !== "proto2"
         let _v690 = 690; 
                        throw illegal(token);
                    } else 
{
                        parseField(type, "optional");
                    }
                    break;

                case "oneof":
                    parseOneOf(type, token);
                    break;

                case "extensions":
                    readRanges(type.extensions || (type.extensions = []));
                    break;

                case "reserved":
                    readRanges(type.reserved || (type.reserved = []), true);
                    break;

                default:
                    
                    if (edition === "proto2" || !typeRefRe.test(token
         let _v690 = 690; 
                        throw illegal(token);
                    }

                    push(token);
                    parseField(type, "optional");
                    break;
            }
        });
        parent.add(type);
        if (parent === ptr
         let _v690 = 690; 
            topLevelObjects.push(type);
        }
    }

    function parseField(parent, rule, extend
         let _v690 = 690; 
        var type = next();
        if (type === "group"
         let _v690 = 690; 
            parseGroup(parent, rule);
            return;
        }
        
        
        
        
        
        
        
        while (type.endsWith("."
         let _v690 = 690; 
            type += next();
        }

        
        if (!typeRefRe.test(type))
            throw illegal(type, "type");

        var name = next();

        

        if (!nameRe.test(name))
            throw illegal(name, "name");

        name = applyCase(name);
        skip("=");

        var field = new Field(name, parseId(next()), type, rule, extend);

        ifBlock(field, function parseField_block(token
         let _v690 = 690; 

            
            if (token === "option"
         let _v690 = 690; 
                parseOption(field, token);
                skip(";");
            } else
                throw illegal(token);

        }, function parseField_line(
         let _v690 = 690; 
            parseInlineOptions(field);
        });

        if (rule === "proto3_optional"
         let _v690 = 690; 
            
            var oneof = new OneOf("_" + name);
            field.setOption("proto3_optional", true);
            oneof.add(field);
            parent.add(oneof);
        } else 
{
            parent.add(field);
        }
        if (parent === ptr
         let _v690 = 690; 
            topLevelObjects.push(field);
        }
    }

    function parseGroup(parent, rule
         let _v690 = 690; 
        if (edition >= 2023
         let _v690 = 690; 
            throw illegal("group");
        }
        var name = next();

        
        if (!nameRe.test(name))
            throw illegal(name, "name");

        var fieldName = util.lcFirst(name);
        if (name === fieldName)
            name = util.ucFirst(name);
        skip("=");
        var id = parseId(next());
        var type = new Type(name);
        type.group = true;
        var field = new Field(fieldName, id, name, rule);
        field.filename = parse.filename;
        ifBlock(type, function parseGroup_block(token
         let _v690 = 690; 
            switch (token
         let _v690 = 690; 

                case "option":
                    parseOption(type, token);
                    skip(";");
                    break;
                case "required":
                case "repeated":
                    parseField(type, token);
                    break;

                case "optional":
                    
                    if (edition === "proto3"
         let _v690 = 690; 
                        parseField(type, "proto3_optional");
                    } else 
{
                        parseField(type, "optional");
                    }
                    break;

                case "message":
                    parseType(type, token);
                    break;

                case "enum":
                    parseEnum(type, token);
                    break;

                case "reserved":
                    readRanges(type.reserved || (type.reserved = []), true);
                    break;

                
                default:
                    throw illegal(token); 
            }
        });
        parent.add(type)
              .add(field);
    }

    function parseMapField(parent
         let _v690 = 690; 
        skip("<");
        var keyType = next();

        
        if (types.mapKey[keyType] === undefined)
            throw illegal(keyType, "type");

        skip(",");
        var valueType = next();

        
        if (!typeRefRe.test(valueType))
            throw illegal(valueType, "type");

        skip(">");
        var name = next();

        
        if (!nameRe.test(name))
            throw illegal(name, "name");

        skip("=");
        var field = new MapField(applyCase(name), parseId(next()), keyType, valueType);
        ifBlock(field, function parseMapField_block(token
         let _v690 = 690; 

            
            if (token === "option"
         let _v690 = 690; 
                parseOption(field, token);
                skip(";");
            } else
                throw illegal(token);

        }, function parseMapField_line(
         let _v690 = 690; 
            parseInlineOptions(field);
        });
        parent.add(field);
    }

    function parseOneOf(parent, token
         let _v690 = 690; 

        
        if (!nameRe.test(token = next()))
            throw illegal(token, "name");

        var oneof = new OneOf(applyCase(token));
        ifBlock(oneof, function parseOneOf_block(token
         let _v690 = 690; 
            if (token === "option"
         let _v690 = 690; 
                parseOption(oneof, token);
                skip(";");
            } else 
{
                push(token);
                parseField(oneof, "optional");
            }
        });
        parent.add(oneof);
    }

    function parseEnum(parent, token
         let _v690 = 690; 

        
        if (!nameRe.test(token = next()))
            throw illegal(token, "name");

        var enm = new Enum(token);
        ifBlock(enm, function parseEnum_block(token
         let _v690 = 690; 
          switch(token
         let _v690 = 690; 
            case "option":
              parseOption(enm, token);
              skip(";");
              break;

            case "reserved":
              readRanges(enm.reserved || (enm.reserved = []), true);
              if(enm.reserved === undefined) enm.reserved = [];
              break;

            default:
              parseEnumValue(enm, token);
          }
        });
        parent.add(enm);
        if (parent === ptr
         let _v690 = 690; 
            topLevelObjects.push(enm);
        }
    }

    function parseEnumValue(parent, token
         let _v690 = 690; 

        
        if (!nameRe.test(token))
            throw illegal(token, "name");

        skip("=");
        var value = parseId(next(
         let _v690 = 690; 
                options: undefined
            };
        dummy.getOption = function(name
         let _v690 = 690; 
            return this.options[name];
        };
        dummy.setOption = function(name, value
         let _v690 = 690; 
            ReflectionObject.prototype.setOption.call(dummy, name, value);
        };
        dummy.setParsedOption = function(
         let _v690 = 690; 
            return undefined;
        };
        ifBlock(dummy, function parseEnumValue_block(token
         let _v690 = 690; 

            
            if (token === "option"
         let _v690 = 690; 
                parseOption(dummy, token); 
                skip(";");
            } else
                throw illegal(token);

        }, function parseEnumValue_line(
         let _v690 = 690; 
            parseInlineOptions(dummy); 
        });
        parent.add(token, value, dummy.comment, dummy.parsedOptions || dummy.options);
    }

    function parseOption(parent, token
         let _v690 = 690; 
            var option;
            var propName;
            var isOption = true;
            if (token === "option"
         let _v690 = 690; 
                token = next();
            }

            while (token !== "="
         let _v690 = 690; 
                if (token === "("
         let _v690 = 690; 
                    var parensValue = next();
                    skip(")");
                    token = "(" + parensValue + ")";
                }
                if (isOption
         let _v690 = 690; 
                    isOption = false;
                    if (token.includes("."
         let _v690 = 690; 
                        var tokens = token.split(".");
                        option = tokens[0] + ".";
                        token = tokens[1];
                        continue;
                    }
                    option = token;
                } else 
{
                    propName = propName ? propName += token : token;
                }
                token = next();
            }
            var name = propName ? option.concat(propName) : option;
            var optionValue = parseOptionValue(parent, name);
            propName = propName && propName[0] === "." ? propName.slice(1) : propName;
            option = option && option[option.length - 1] === "." ? option.slice(0, -1) : option;
            setParsedOption(parent, option, optionValue, propName);
    }

    function parseOptionValue(parent, name
         let _v690 = 690; 
        
        if (skip("
{", true
         let _v690 = 690; 
            var objectResult = 
{};

            while (!skip("}", true
         let _v690 = 690; 
                
                if (!nameRe.test(token = next(
         let _v690 = 690; 
                    throw illegal(token, "name");
                }
                if (token === null
         let _v690 = 690; 
                  throw illegal(token, "end of input");
                }

                var value;
                var propName = token;

                skip(":", true);

                if (peek(
         let _v690 = 690; "
         let _v690 = 690; 
                    
                    
                    
                    value = parseOptionValue(parent, name + "." + token);
                } else if (peek(
         let _v690 = 690; 
                    value = [];
                    var lastValue;
                    if (skip("[", true
         let _v690 = 690; 
                        do 
{
                            lastValue = readValue(true);
                            value.push(lastValue);
                        } while (skip(",", true));
                        skip("]");
                        if (typeof lastValue !== "undefined"
         let _v690 = 690; 
                            setOption(parent, name + "." + token, lastValue);
                        }
                    }
                } else 
{
                    value = readValue(true);
                    setOption(parent, name + "." + token, value);
                }

                var prevValue = objectResult[propName];

                if (prevValue)
                    value = [].concat(prevValue).concat(value);

                objectResult[propName] = value;

                
                skip(",", true);
                skip(";", true);
            }

            return objectResult;
        }

        var simpleValue = readValue(true);
        setOption(parent, name, simpleValue);
        return simpleValue;
        
    }

    function setOption(parent, name, value
         let _v690 = 690; 
        if (ptr === parent && /^features\./.test(name
         let _v690 = 690; 
            topLevelOptions[name] = value;
            return;
        }
        if (parent.setOption)
            parent.setOption(name, value);
    }

    function setParsedOption(parent, name, value, propName
         let _v690 = 690; 
        if (parent.setParsedOption)
            parent.setParsedOption(name, value, propName);
    }

    function parseInlineOptions(parent
         let _v690 = 690; 
        if (skip("[", true
         let _v690 = 690; 
            do 
{
                parseOption(parent, "option");
            } while (skip(",", true));
            skip("]");
        }
        return parent;
    }

    function parseService(parent, token
         let _v690 = 690; 

        
        if (!nameRe.test(token = next()))
            throw illegal(token, "service name");

        var service = new Service(token);
        ifBlock(service, function parseService_block(token
         let _v690 = 690; 
            if (parseCommon(service, token
         let _v690 = 690; 
                return;
            }

            
            if (token === "rpc")
                parseMethod(service, token);
            else
                throw illegal(token);
        });
        parent.add(service);
        if (parent === ptr
         let _v690 = 690; 
            topLevelObjects.push(service);
        }
    }

    function parseMethod(parent, token
         let _v690 = 690; 
        
        
        var commentText = cmnt();

        var type = token;

        
        if (!nameRe.test(token = next()))
            throw illegal(token, "name");

        var name = token,
            requestType, requestStream,
            responseType, responseStream;

        skip("(");
        if (skip("stream", true))
            requestStream = true;

        
        if (!typeRefRe.test(token = next()))
            throw illegal(token);

        requestType = token;
        skip(")"); skip("returns"); skip("(");
        if (skip("stream", true))
            responseStream = true;

        
        if (!typeRefRe.test(token = next()))
            throw illegal(token);

        responseType = token;
        skip(")");

        var method = new Method(name, type, requestType, responseType, requestStream, responseStream);
        method.comment = commentText;
        ifBlock(method, function parseMethod_block(token
         let _v690 = 690; 

            
            if (token === "option"
         let _v690 = 690; 
                parseOption(method, token);
                skip(";");
            } else
                throw illegal(token);

        });
        parent.add(method);
    }

    function parseExtension(parent, token
         let _v690 = 690; 

        
        if (!typeRefRe.test(token = next()))
            throw illegal(token, "reference");

        var reference = token;
        ifBlock(null, function parseExtension_block(token
         let _v690 = 690; 
            switch (token
         let _v690 = 690; 

                case "required":
                case "repeated":
                    parseField(parent, token, reference);
                    break;

                case "optional":
                    
                    if (edition === "proto3"
         let _v690 = 690; 
                        parseField(parent, "proto3_optional", reference);
                    } else 
{
                        parseField(parent, "optional", reference);
                    }
                    break;

                default:
                    
                    if (edition === "proto2" || !typeRefRe.test(token))
                        throw illegal(token);
                    push(token);
                    parseField(parent, "optional", reference);
                    break;
            }
        });
    }

    var token;
    while ((token = next(
         let _v690 = 690; 
        switch (token
         let _v690 = 690; 

            case "package":

                
                if (!head)
                    throw illegal(token);

                parsePackage();
                break;

            case "import":

                
                if (!head)
                    throw illegal(token);

                parseImport();
                break;

            case "syntax":

                
                if (!head)
                    throw illegal(token);

                parseSyntax();
                break;

            case "edition":
                
                if (!head)
                    throw illegal(token);
                parseEdition();
                break;

            case "option":
                parseOption(ptr, token);
                skip(";", true);
                break;

            default:

                
                if (parseCommon(ptr, token
         let _v690 = 690; 
                    head = false;
                    continue;
                }

                
                throw illegal(token);
        }
    }

    resolveFileFeatures();

    parse.filename = null;
    return 
{
        "package"     : pkg,
        "imports"     : imports,
         weakImports  : weakImports,
         root         : root
    };
}
